synapse/lib/agenda.py

```diff
@@ -0,0 +1,614 @@
+            daysinmonth = calendar.monthrange(newyear, newmonth)[1]
+            dayofmonthreq = reqdict.get(TimeUnit.DAYOFMONTH)
+            if dayofmonthreq is not None:
+                newday = _dayofmonth(dayofmonthreq, newmonth, newyear)
+            else:
+                newday = min(daysinmonth, dt.day)
+            return dt.replace(day=newday, month=newmonth, year=newyear)
+        if incunit == TimeUnit.DAY:
+            return dt + datetime.timedelta(days=incval)
+        if incunit == TimeUnit.DAYOFWEEK:
+            # incval in this case means next day of week whose weekday matches incval (0-6)
+            days = (6 + incval - dt.weekday()) % 7 + 1
+            newdt = dt + datetime.timedelta(days=days)
+            assert newdt.weekday() == incval
+            return newdt
+        if incunit == TimeUnit.HOUR:
+            return dt + datetime.timedelta(hours=incval)
+        if incunit == TimeUnit.MINUTE:
+            return dt + datetime.timedelta(minutes=incval)
+        else:
+            assert 0, 'Invalid incunit'  # pragma: no cover
+
+class _Appt:
+    '''
+    A single entry in the Agenda:  a storm query to run in the future, potentially more than once
+
+    Each such entry has a list of ApptRecs.  Each time the appointment is scheduled, the nexttime of the appointment is
+    the lowest nexttime of all its ApptRecs.
+    '''
+    def __init__(self, iden, recur, indx, query, username, recs, nexttime=None):
+        self.iden = iden
+        self.recur = recur # does this appointment repeat
+        self.indx = indx  # incremented for each appt added ever.  Used for nexttime tiebreaking for stable ordering
+        self.query = query  # query to run
+        self.username = username  # user to run query as
+        self.recs = recs  # List[ApptRec]
+        self._recidxnexttime = None # index of rec who is up next
+
+        if self.recur and not self.recs:
+            raise s_exc.BadTime(mesg='A recurrent appointment with no records')
+
+        if nexttime is None and self.recs:
+            now = time.time()
+            self.nexttime = now
+            self.updateNexttime(now + 1.0)  # lie slightly about the time so it does advance
+            if self.nexttime is None:
+                raise s_exc.BadTime(mesg='Appointment is in the past')
+        else:
+            self.nexttime = nexttime
+        self.isrunning = False  # whether it is currently running
+        self.startcount = 0  # how many times query has started
+        self.laststarttime = None
+        self.lastfinishtime = None
+        self.lastresult = None
+        self.enabled = True
+
+    def __eq__(self, other):
+        ''' For heap logic to sort upcoming events lower '''
+        return (self.nexttime, self.indx) == (other.nexttime, other.indx)
+
+    def __lt__(self, other):
+        ''' For heap logic '''
+        return (self.nexttime, self.indx) < (other.nexttime, other.indx)
+
+    def todict(self):
+        return {
```

Comment:
comment

Reply:
we've previously used `pack()` to serialize objects instead of `todict()`

Reply:
ok